src/gencost/waterfall.py

```diff
@@ -2232,27 +2366,485 @@ def valid_generation(df_):
+        xwalk = self.xwalk
+
+        return (
+            self.get_exa_by_generator()
+            .merge(
+                xwalk[["plant_id_eia", "generator_id", "prime_mover", "fuel_group"]],
+                on=["plant_id_eia", "generator_id", "prime_mover"],
+                how="left",
+            )
+            .assign(
+                age=lambda x: (
+                    ((pd.datetime.now() - x.generator_operating_date).dt.days) / 365.25
+                ).round(2),
+                report_year=lambda x: x.report_date.dt.year,
+            )
+            # .assign(year=lambda x: x["report_date"].dt.year)
+            .pipe(self.create_fill_in_ep_thresholds)
+        )
+
+    def fill_in_ep_data(self):
+        ep = self.get_ep_data()
+
+        missing_data = (
+            self.find_missing_data()
+            .pipe(self.create_fill_in_ep_thresholds)
+            .assign(
+                essentials_present=lambda x: np.where(
+                    x["essentials"].isin(ep["essentials"]),
+                    "essentials",
+                    pd.NA,
+                ),
+                ba_plus_essentials_present=lambda x: np.where(
+                    x["ba_plus_essentials"].isin(ep["ba_plus_essentials"]),
+                    "ba_plus_essentials",
+                    pd.NA,
+                ),
+                ba_plus_age_present=lambda x: np.where(
+                    x["ba_plus_age"].isin(ep["ba_plus_age"]),
+                    "ba_plus_age",
+                    pd.NA,
+                ),
+            )  # make an exception of when there is no ba code or age
+            .assign(
+                ba_plus_age_present=lambda x: np.where(
+                    (x["final_ba_code"].isnull()) | (x["age"].isnull()),
+                    pd.NA,
+                    x["ba_plus_age_present"],
+                ),
+                ba_plus_essentials_present=lambda x: np.where(
+                    (x["final_ba_code"].isnull()),
+                    pd.NA,
+                    x["ba_plus_essentials_present"],
+                ),
+                # fill_in_score=lambda x: np.where(
+                # x["ba_plus_age_present"].notnull(), x["ba_plus_age_present"]
+                # ),
+                match=lambda x: x["ba_plus_age_present"]
+                .fillna(x["ba_plus_essentials_present"])
+                .fillna(x["essentials_present"]),
+            )
+        )
+
+        """
+        Merge missing data df with historical data
+        1) Loop based on unique values in match column
+        2) Query based on value (list with different kind of matches)
+        3) Merge
+        4) Append to a list
+        5) drop duplicates since they're might be multiple matches
+
+        """
+
+        cols = ["ba_plus_age", "ba_plus_essentials", "essentials"]
+        filled_in = []
+
+        for col in cols:
+            # core columns we want to get from historical, append column we're going to merge on
+            HIST_EP_COLS.append(col)
```

Comment:
I might not be understanding this process correctly but I think there are some columns in this list that I don't think we want to fill in using historical values for similar generators. What we really want is to cast the generator's current value to previous years. On this list of columns I think are the following:

    "capacity_mw",
    "associated_combined_heat_power",
    "duct_burners",
    "bypass_heat_recovery",
    "solid_fuel_gasification",
    "carbon_capture",
    "fluidized_bed_tech",
    "pulverized_coal_tech",
    "stoker_tech",
    "other_combustion_tech",
    "subcritical_tech",
    "supercritical_tech",
    "ultrasupercritical_tech",
    "real_pollution_control_costs_per_kw",

Then there are some that may need to get recalculated, if not now, when we cast the eternal present data into the future, though that will raise problems for the inflation related ones.

    "age_in_report_year",
    "age_in_current_year",
    "age_of_observation",
    "age_relative_to_prime_avg",
    "wage_scale",
    "age_of_observation_secular_adj",
    "inflator_to_2021",

The straggler is "pollution_control_costs_per_kw", I think the right way to do this is to calculate it from  "real_pollution_control_costs_per_kw" by dividing by the inflator.

Reply:
@arengel, should I take this to mean that there are three different ways to fill in data?

This is my interpretation of what that might look like:
For the generation and fuel consumption columns: fill in historical data from similar generator
For capacity, tech, and pollution control costs: fill in data from current year to all previous years
Age, wage, and inflation: re-calculate using report date and generator operating date columns

Reply:
Exactly. Though as I think about it more, let's leave the following columns in the first group that you fill from historical data and I will create an issue to address a better resolution for them:

    "wage_scale",
    "age_of_observation_secular_adj",
    "inflator_to_2021",